README.md
# libeXaDrums

 ExaDrums is a software drum module that allows drummers play with
 custom-made drum kits.
 This C++ library provides the most common features of a drum module.
 Its associated graphical user interface can be found in a separate
 repository: <https://github.com/SpintroniK/eXaDrums.>

 Each drum kit is made of instruments that can be individually
 controlled.

 A built-in metronome can be combined with a rhythm coach to make
 practice sessions easier and efficient.

 The drum triggers can be adjusted so that their response feels as
 natural as possible, and different sensor interfaces include
 external sensors.

 Although eXaDrums is usable as a drum module, it is still a
 young project. As such, some features are not yet implemented
 or are still experimental.

## Table of content

- [Installation and configuration](#installation-and-configuration)
  - [Dependencies](#dependencies)
  - [Building libeXaDrums](#building-libexadrums)
  - [Installation](#installation)
  - [Configuration](#configuration)
- [Usage](#usage)

## Installation and configuration

Libexadrums is available from the Debian unstable and testing distributions, as well as Ubuntu >=19.10.
If you wish to install the latest version of the library, you can build the packages from the Github repositorie.

### Dependencies

libeXaDrums depends on Alsa, TinyXml2 and minizip. You can install those two libraries using this command line:

```shell
sudo apt install libasound2-dev libtinyxml2-dev libminizip-dev
```

LibeXaDrums also depends on build-essential, autotools and pkg-config in order to build and install the binaries:

```shell
sudo apt install autoconf automake libtool build-essential pkg-config
```

If you wish to build the Debian packages (which is the recommended way to install
the library), you will need to install debhelper:

```shell
sudo apt install debhelper
```

You will also need to install git, so that you can clone this repository:

```shell
sudo apt install git
git clone https://github.com/SpintroniK/libeXaDrums.git
```

### Building libeXaDrums

Now that you have all dependencies, you can build the Debian packages.
Make sure you have dehelper installed, and then checkout the debian branch
from your cloned repository:

```shell
cd libeXaDrums
git checkout debian
```

Then you can build the package:

```shell
dpkg-buildpackage -b -uc -us
```

If you have multiple cores/threads, you can speed up the build process by appending the option -jn to dpkg-buildpackage, where n is the number of threads that you want to use for the compilation.
Example, for four threads type: `dpkg-buildpackage -b -uc -us -j4`.

### Installation

The Debian packages are built in the parent directory, so you should be able to install them by using dpkg (don't forget to install them as root):

```shell
cd ..
sudo apt install ./libexadrums0_[...].deb
sudo apt install ./libexadrums-dev[...].deb
```

### Configuration

## Usage

The library is self documented.


debian/changelog
libexadrums (0.7.0-1) unstable; urgency=medium

  * New upstream release.
    Change Shared Object Version in library package name.
  * Update standards version to 4.6.2, no changes needed.
  * Set upstream metadata fields: Bug-Submit, Repository.
  * Fix watch file.

 -- Jeremy Oden <jerem.oden@gmail.com>  Tue, 20 Dec 2022 02:32:43 +0000

libexadrums (0.6.0-1) unstable; urgency=medium

  * New upstream version.
  * Change homepage URL.

 -- Jeremy Oden <jerem.oden@gmail.com>  Tue, 29 Dec 2020 17:51:59 +0100

libexadrums (0.5.2-2) unstable; urgency=medium

  * Source only upload for migration to testing.

 -- Jeremy Oden <jerem.oden@gmail.com>  Sun, 06 Dec 2020 15:49:07 +0100

libexadrums (0.5.2-1) unstable; urgency=medium

  * New upstream version.
    Rename the library version because of the new Shared Object version.
    New dependency on libminizip.
  * Depend on zlib1g-dev because libminizip-dev does not.
  * Select Debian branch in Vcs-Browser.
  * Swap binary packages in control files in order to avoid hard-coding
    the SO version in the packaging.
  * Debhelper 13.  Refine patterns in debhelper config files.
  * Drop obsolete work-around for libtool bug.
  * Mark autopkgtests as superficial.
  * Add upstream/metadata.
  * Standards-Version 4.5.1 (no changes)

 -- Jeremy Oden <jerem.oden@gmail.com>  Sat, 28 Nov 2020 17:25:26 +0100

libexadrums (0.4.2-1) unstable; urgency=medium

  * New upstream version.

 -- Jeremy Oden <jerem.oden@gmail.com>  Sun, 06 Oct 2019 17:46:53 +0100

libexadrums (0.4.1-1) unstable; urgency=medium

  * New upstream version.
  * Added 'Multi-Arch: same' to libexadrums-dev.
  * Package libexadrums-dev depends on libasound-dev. Closes: #933173.

  [ Nicolas Boulenguez ]
  * Debhelper 12.

 -- Jeremy Oden <jerem.oden@gmail.com>  Thu, 08 Aug 2019 18:53:45 +0100

libexadrums (0.4.0-1) unstable; urgency=medium

  * New upstream version.
  * First upload to unstable.
  * VCS-Git: select packaging branch.
  * Standards-Version: 4.4.0.

 -- Jeremy Oden <jerem.oden@gmail.com>  Tue, 09 Jul 2019 16:19:10 +0200

libexadrums (0.3.0-2) experimental; urgency=medium

  * Add -pthread -latomic --no-undefined build flags. Closes: #927229.

  [ Nicolas Boulenguez ]
  * Move --as-needed before libraries it is supposed to act on.

 -- Jeremy Oden <jerem.oden@gmail.com>  Sat, 27 Apr 2019 11:46:50 +0200

libexadrums (0.3.0-1) experimental; urgency=medium

  * Initial Debian packaging. Closes: #919327.

 -- Jeremy Oden <jerem.oden@gmail.com>  Fri, 01 Feb 2019 20:24:25 +0000


Util/Crypt.h
/*
 * Crypt.h
 *
 *  Created on: 26 Feb 2018
 *      Author: jeremy
 */

#ifndef SOURCE_UTIL_CRYPT_H_
#define SOURCE_UTIL_CRYPT_H_

#include <array>
#include <vector>
#include <string>
#include <algorithm>

namespace Util
{

	static constexpr char Base64Chars[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

	static constexpr std::array<int, 256> Base64Indices = {{ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
		0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
		0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 62, 63, 62, 62, 63, 52, 53, 54, 55,
	   56, 57, 58, 59, 60, 61,  0,  0,  0,  0,  0,  0,  0,  0,  1,  2,  3,  4,  5,  6,
		7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,  0,
		0,  0,  0, 63,  0, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
	   41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51 }};

	static std::string Base64Encode(const std::vector<uint8_t>& str)
	{
		const auto len = str.size();
		size_t d = len % 3;

		std::string str64(4 * (int(d > 0) + len / 3), '=');

		for(size_t i = 0, j = 0; i < len - d; i += 3)
		{
			int32_t n = int32_t(str[i]) << 16 | int32_t(str[i + 1]) << 8 | str[i + 2];
			str64[j++] = Base64Chars[n >> 18];
			str64[j++] = Base64Chars[n >> 12 & 0x3F];
			str64[j++] = Base64Chars[n >> 6 & 0x3F];
			str64[j++] = Base64Chars[n & 0x3F];
		}

		if(d-- > 0)	// Padding
		{
			int32_t n = d ? int32_t(str[len - 2]) << 8 | str[len - 1] : str[len - 1];
			str64[str64.size() - 2] = d ? Base64Chars[(n & 0xF) << 2] : '=';
			str64[str64.size() - 3] = d ? Base64Chars[n >> 4 & 0x03F] : Base64Chars[(n & 3) << 4];
			str64[str64.size() - 4] = d ? Base64Chars[n >> 10] : Base64Chars[n >> 2];
		}

		return str64;
	}

	template <typename T>
	static std::string Base64Encode(const std::vector<T>& input)
	{
	    std::vector<uint8_t> str;
	    str.reserve(sizeof(T) * input.size());

	    std::copy_n(reinterpret_cast<const char*>(input.data()), sizeof(T) * input.size(), std::back_inserter(str));

	    return Base64Encode(str);
	}

	/*static std::string Base64Decode(const std::string& str64)
	{
		const auto len = str64.size();
		const auto pad = len > 0 && (len % 4 || str64[len - 1] == '=');
		const size_t length = ((len + 3) / 4 - pad) * 4;

		std::string str(length / 4 * 3 + pad, '\0');

		for(size_t i = 0, j = 0; i < length; i += 4)
		{
			int32_t n = Base64Indices[str64[i]] << 18 | Base64Indices[str64[i + 1]] << 12 | Base64Indices[str64[i + 2]] << 6 | Base64Indices[str64[i + 3]];
			str[j++] = n >> 16;
			str[j++] = n >> 8 & 0xFF;
			str[j++] = n & 0xFF;
		}

		if(pad != 0)
		{
			int32_t n = Base64Indices[str64[length]] << 18 | Base64Indices[str64[length + 1]] << 12;
			str.back() = n >> 16;

			if (len > length + 2 && str64[length + 2] != '=')
			{
				n |= Base64Indices[str64[length + 2]] << 6;
				str.push_back(n >> 8 & 0xFF);
			}
		}

		return str;
	}
	*/
}

#endif /* SOURCE_UTIL_CRYPT_H_ */
